### DISCUSSION TITLE
Socket IO - NextJS - {target: "server"}

### DISCUSSION THREAD

**Discussion by ShanonJackson** [vercel/vercel#4453] [Help] [answered]
Any  ideas why this works locally but NOT on production publishing to vercel? i have no now.json changing any configuration? Socket connection just returns 404, however npm run build && npm start locally works completely fine, i tried using port 80 in production still 404.
```
//    ./server.js
const app = require("next")({ dev: process.env.NODE_ENV !== "production" });
const handler = app.getRequestHandler();
const { createServer }  = require("http");
const { parse } = require("url");
const Kafka = require("./kafka").ConnectKafka;

app.prepare().then(() => {
	const server = createServer((req, res) => {
		return handler(req, res, parse(req.url, true), io)
	});
	const io = require("socket.io")(server);
	io.on("connection", (socket) => {
		// doesn't matter.
	});
	server.listen(3000, (err) => {
		if (err) throw err;
		console.log("> Ready on http://localhost:3000");
	});
});
```
and package.json scripts
```
    "build": "next build",
    "start": "cross-env NODE_ENV=production node server.js",
```

**Comment by coetry** [comment, marked as answer]
Hi @ShanonJackson we don't support WebSockets since they are long-lived TCP connections that don't fit the constraints of a [Serverless Function](https://vercel.com/docs/v2/serverless-functions/introduction#) environment. 

If you'd like to use WebSockets for real-time communication on your Vercel projects, I would recommend a mature solution like [Pusher](https://pusher.com/) which handles the WebSocket infrastracture for you and also provides a great developer experience.
